[feat] Add import failure detection and error handling for package manager

viva-jinyi opened the pull request:
## Summary
- Implements comprehensive import failure detection across Manager components
- Provides clear visual feedback when packages fail to import
- Adds detailed error reporting through clickable error badges

## Changes
- **Import failure detection**: Added `importFailed` computed properties to InfoPanel, PackCardFooter, and WarningTabPanel
- **UI updates**: Hide install/enable buttons when import failed, show error badges instead
- **Error reporting**: Clickable error badges open detailed import failure messages in modal dialogs
- **Status handling**: Update PackStatusMessage to display "Install Error" status with error styling
- **Logs access**: Add "Open logs" button in WarningTabPanel for import_failed conflicts
- **Message utilities**: Extend conflictMessageUtil to handle import_failed conflict type
- **Translations**: Add new i18n keys for import failure messages and error states

## Test plan
- [ ] Test with packages that have import_failed conflicts
- [ ] Verify error badges appear instead of install/enable buttons
- [ ] Confirm clickable error badges show detailed error messages
- [ ] Check "Open logs" button functionality in warning panel
- [ ] Validate proper error styling and translations

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4600-feat-Add-import-failure-detection-and-error-handling-for-package-manager-2406d73d365081958797dedb6db920fa) by [Unito](https://www.unito.io)

viva-jinyi committed:
[feat] Add import failure detection and error handling for package manager

- Add importFailed detection across InfoPanel, PackCardFooter, and WarningTabPanel components
- Hide install/enable buttons when import failed, show error badge instead
- Add clickable error badges that display detailed import failure messages
- Update PackStatusMessage to handle importFailed status with error styling
- Add Open logs button in WarningTabPanel for import_failed conflicts
- Extend conflictMessageUtil to support import_failed conflict type
- Add new translation keys for import failure messages and error states

christian-byrne commented on src/components/dialog/content/manager/infoPanel/tabs/WarningTabPanel.vue:
```diff
@@ -1,24 +1,66 @@
+      @click="showImportFailedDialog"
+    >
+      <i class="pi pi-code text-base"></i>
+      <span class="text-white text-sm">{{ t('serverStart.openLogs') }}</span>
```
If setting text color explicitly, make sure to switch colors based on light/dark with `dark-theme:` prefix.

christian-byrne commented on src/components/dialog/content/manager/packCard/PackCardFooter.vue:
```diff
@@ -38,23 +52,56 @@ const { nodePack } = defineProps<{
 
 const { getConflictsForPackageByID } = useConflictDetectionStore()
 const { checkNodeCompatibility } = useConflictDetection()
+const { showErrorDialog } = useDialogService()
+
+const conflicts = computed(() => getConflictsForPackageByID(nodePack.id!))
+
+const importFailedInfo = computed(() => {
+  if (!nodePack.id || !conflicts.value) return null
+
+  const importFailedConflicts = conflicts.value.conflicts.filter(
+    (item) => item.type === 'import_failed'
+  )
+
+  return importFailedConflicts.length > 0 ? importFailedConflicts : null
+})
+
+const importFailed = computed(() => {
+  return importFailedInfo.value !== null
+})
+
+const showImportFailedDialog = () => {
+  if (importFailedInfo.value) {
+    const errorMessage =
+      importFailedInfo.value
+        .map((conflict) => conflict.required_value)
+        .filter(Boolean)
+        .join('\n') || t('manager.importFailedGenericError')
+
+    const error = new Error(errorMessage)
+
+    showErrorDialog(error, {
+      title: t('manager.failedToInstall'),
+      reportType: 'importFailedError'
+    })
+  }
+}
```
Since we use this logic in 3 places, how about moving it to its own composable?

viva-jinyi replied:
We can just use provide/inject for the shared logic between components, and use a composable specifically for the importFailed-related logic. That should keep things clean and organized.

christian-byrne reviewed (commented):
Suggestion: we may be able to avoid prop drilling a bit by using https://vuejs.org/guide/components/provide-inject#prop-drilling. The providers could be InfoPanel and PackCard.

What do you think?

viva-jinyi committed:
[refactor] Simplify ternary expressions and fix dark theme text color

- Simplify complex nested ternary expressions in PackStatusMessage using if statements
- Fix missing dark theme text color in WarningTabPanel "Open logs" button

viva-jinyi committed:
[refactor] Simplify import failed detection with unified composable

- Create useImportFailedDetection composable to centralize import failed logic
- Use provide/inject pattern to eliminate prop drilling in InfoPanel components
- Simplify composable to accept only packageId parameter for better maintainability
- Remove duplicate import failed detection code across PackCardFooter and WarningTabPanel
- Add ImportFailedContext type for consistent provide/inject usage
- Fix reactivity issues with computed refs in component communication

Components using inject pattern:
- InfoPanelHeader, PackStatusMessage, InfoTabs

Components using composable directly:
- PackCardFooter, WarningTabPanel (via packageId)

viva-jinyi committed:
[test] Add comprehensive unit tests for useImportFailedDetection composable

- Add 9 unit tests covering all core functionality and edge cases
- Test import failed detection logic for installed vs non-installed packages
- Test filtering of import_failed conflict types vs other conflict types
- Test reactive behavior with computed ref packageId parameter
- Test error dialog integration with proper error message handling
- Test edge cases with null/undefined packageId inputs
- Mock external dependencies (stores, services, i18n) properly for isolated testing
- Achieve full test coverage for the new composable

All tests pass (9/9) ensuring robust functionality and preventing regressions.

viva-jinyi commented:
### Summary of Changes:
  - Created useImportFailedDetection composable to centralize import failure detection logic and eliminate code duplication
  - Implemented provide/inject pattern to solve prop drilling issues across InfoPanel components
  - Simplified composable API to accept only packageId parameter for better maintainability and clarity
  - Fixed reactivity issues with proper computed ref handling for dynamic package selection
  - Added comprehensive test coverage with 9 unit tests covering all functionality and edge cases

### Key Technical Improvements:
  - ✅ Code Deduplication: Removed duplicate import failed logic from PackCardFooter and WarningTabPanel
  - ✅ Better Architecture: Centralized logic in reusable composable with consistent API
  - ✅ Prop Drilling Elimination: Used provide/inject for ImportFailedContext in InfoPanel component tree
  - ✅ Reactivity Fix: Proper handling of computed refs for dynamic package ID changes
  - ✅ Type Safety: Added ImportFailedContext interface and ImportFailedKey injection key

### Components Updated:
  - Using inject pattern: InfoPanelHeader, PackStatusMessage, InfoTabs
  - Using composable directly: PackCardFooter, WarningTabPanel (via packageId)
  - Providing context: InfoPanel (root provider)

### Test Coverage:
  - ✅ 9/9 unit tests passing for the new composable
  - ✅ All existing tests maintained (951 total tests, 919 passing)
  - ✅ Edge case coverage: null/undefined inputs, reactive behavior, error dialog integration

christian-byrne reviewed (approved):
LGTM!

viva-jinyi committed:
[feature] Add conditional description display in NodeConflictDialogContent

- Add `showAfterWhatsNew` prop to NodeConflictDialogContent to control description visibility
- Show description section only when conflict modal appears after What's New dismissal
- Update SidebarHelpCenterIcon to pass showAfterWhatsNew: true for post-update conflicts
- Improve UX by hiding redundant description text in other conflict modal scenarios

This ensures the description only appears in the specific case where users see conflicts
after ComfyUI updates, avoiding unnecessary text in regular conflict scenarios.

viva-jinyi committed:
[style] max-height deleted

viva-jinyi commented on src/composables/useImportFailedDetection.ts:
@christian-byrne How about putting it in a manager folder instead?
Or we could even create a dedicated conflict folder and organize all related composables there — I think that might make things clearer.

viva-jinyi replied:
Alright, I think it would be good to refactor it once more. It might help to separate concerns by reorganizing the folder structure or renaming some files.

viva-jinyi committed:
[refactor] Optimize NodeConflictDialogContent to use composables directly

- Remove props drilling by using useConflictDetection composable directly
- Add Import Failed Extensions section at the top of conflict dialog
- Show package names for import failed conflicts
- Optimize data extraction with lodash for better performance
- Update dialog service to support minimal props passing
- Add translation key for importFailedExtensions

viva-jinyi committed:
[fix] deleted hard coding

viva-jinyi committed:
[test] Add comprehensive tests for NodeConflictDialogContent

- Add tests for import failed extensions section display
- Test panel interactions and toggling behavior
- Test empty states and data filtering with lodash
- Mock useConflictDetection composable for isolated testing
- Verify correct separation of import_failed conflicts
- Test accessibility and scrolling behavior
